bcm2835-codec tweaks

6by9 opened the pull request.

6by9 committed:
media/v4l2_m2m: In buffered mode run jobs if either port is streaming

In order to get the intended behaviour of the stateful video
decoder API where only the OUTPUT queue needs to be enabled and fed
buffers in order to get the SOURCE_CHANGED event that configures the
CAPTURE queue, we want the device to run should either queue be
streaming.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging/bcm2835-codec: Correct logging of size_t to %zu

Fixes: "staging/bcm2835-codec: Log the number of excess supported formats"
Which used %u for printing a size_t, and 64bit builds then log a warning.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging/bcm2835-codec: Add support for pixel aspect ratio

If the format is detected by the driver and a V4L2_EVENT_SOURCE_CHANGE
event is generated, then pass on the pixel aspect ratio as well.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
@jc-kynesim I did go for the framework tweak instead of totally reworking the buffer feed code.

@popcornmix Hopefully pixel aspect ratio reporting.

pelwell commented on drivers/media/v4l2-core/v4l2-mem2mem.c:
```diff
@@ -301,8 +301,9 @@ static void __v4l2_m2m_try_queue(struct v4l2_m2m_dev *m2m_dev,
-	    || !m2m_ctx->cap_q_ctx.q.streaming) {
+	if (!m2m_ctx->out_q_ctx.buffered &&
+	    (!m2m_ctx->out_q_ctx.q.streaming ||
+	     !m2m_ctx->cap_q_ctx.q.streaming)) {
```
I may have misunderstood the intention of this commit, but it appears to bypass the streaming check if running in buffered mode.

6by9 replied:
You're right, but it wasn't quite my intent. Too much inverted logic.

I wanted the existing logic when not buffering, but to not return should either queue be streaming when buffered. One to rework.

pelwell commented on drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c:
```diff
@@ -1331,6 +1331,7 @@ static int vidioc_s_fmt(struct bcm2835_codec_ctx *ctx, struct v4l2_format *f,
 	struct vb2_queue *vq;
```
Typo in the commit message - `aat`.

6by9 replied:
Corrected.

pelwell commented on drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c:
```diff
@@ -2464,22 +2545,25 @@ static int bcm2835_codec_start_streaming(struct vb2_queue *q,
 		if (ret)
 			v4l2_err(&ctx->dev->v4l2_dev, "%s: Failed enabling i/p port, ret %d\n",
 				 __func__, ret);
+
```
Is this accidental?

6by9 replied:
Yes. Fixed.

pelwell commented on drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c:
```diff
@@ -2464,22 +2545,25 @@ static int bcm2835_codec_start_streaming(struct vb2_queue *q,
+					 __func__, ret);
+		}
 	}
+	v4l2_err(&ctx->dev->v4l2_dev, "%s: Done, ret %d\n",
```
This looks like an unconditional error message.

6by9 replied:
Sorted.

pelwell commented on drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c:
```diff
@@ -1540,6 +1544,14 @@ static int vidioc_g_selection(struct file *file, void *priv,
 			s->r.width = q_data->crop_width;
 			s->r.height = q_data->crop_height;
 			break;
+		case V4L2_SEL_TGT_CROP_BOUNDS:
```
This change seems unrelated.

6by9 replied:
Yes and no. I'll make it a second patch.

The only route for userspace to query the pixel aspect ratio is via ```VIDIOC_CROPCAP```, which goes through [v4l2_cropcap](https://elixir.bootlin.com/linux/latest/source/drivers/media/v4l2-core/v4l2-ioctl.c#L2451).
That calls our vidioc_g_pixelaspect function, but then follows it up with calls to g_selection for other properties, including these.
Where and how they are relevant for a video decoder is another questsion, but the call fails otherwise.

Oddly the only other decoder I can find which implements vidioc_g_pixelaspect is staging/media/meson/vdec, but it doesn't implement g_selection at all, so I don't believe there is any driver in mainline that actually implements this, but FFmpeg tries to read it.

pelwell reviewed (commented):
You lost me in the second commit, but here's some feedback.

6by9 committed:
staging/bcm2835-codec: Implement additional g_selection calls for decode

v4l_cropcap calls our vidioc_g_pixelaspect function to get the pixel
aspect ratio, but also calls g_selection for V4L2_SEL_TGT_CROP_BOUNDS
and V4L2_SEL_TGT_CROP_DEFAULT. Whilst it allows for vidioc_g_pixelaspect
not to be implemented, it doesn't allow for either of the other two.

Add in support for the additional selection targets.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
Updated to hopefully address your comments.

popcornmix commented:
Kodi doesn't see the aspect ratio, but it might be getting lost in ffmpeg or kodi itself. I'll see if I can find it...

6by9 commented:
Does Kodi / FFmpeg actually make use of the resolution changed event? If it has configured the output correctly then there is no event and we have no chance to report the PAR.

pelwell commented on drivers/media/v4l2-core/v4l2-mem2mem.c:
```diff
@@ -301,9 +301,10 @@ static void __v4l2_m2m_try_queue(struct v4l2_m2m_dev *m2m_dev,
-		dprintk("Streaming needs to be on for both queues\n");
+	if ((!m2m_ctx->out_q_ctx.q.streaming ||
+	     !m2m_ctx->cap_q_ctx.q.streaming) &&
+	    !(m2m_ctx->out_q_ctx.buffered && m2m_ctx->out_q_ctx.q.streaming)) {
```
I just spent a minute or two writing out your condition as you described it:
```
	if (!((m2m_ctx->out_q_ctx.q.streaming &&
	       m2m_ctx->cap_q_ctx.q.streaming) ||
	      (m2m_ctx->out_q_ctx.buffered && m2m_ctx->out_q_ctx.q.streaming))) {
```
and then using De Morgan's laws to transform it into what you have written. But I shouldn't have to - what's wrong with the literal version above?

bitbashing replied:
The OP's version is just an evolution of pre-existing (and thus assumed stable to some degree through testing/deployment) code using a minimal set of changes.

The check could also be written as

```
	if (!(m2m_ctx->out_q_ctx.q.streaming && m2m_ctx-cap_q_ctx.q.streaming) && 
	       !(m2m_ctx->out_q_ctx.buffered && m2m_ctx->out_q_ctx.q.streaming)) {
```

They are all logically equivalent but having OR inside the clause and top-level NOT in the proposed transformed version depending on the conditions could make it a slower performing check if no short-circuiting happens inside and the result is not final until after the negation is performed at the end.

The changes suggested in c54a951 still employ a short-circuit, so they look ok to me.

pelwell replied:
They aren't wrong, but in (especially) kernel work clarity trumps pretty much everything, especially when the compiler will likely generate the same code in both cases.

bitbashing replied:
I agree on clarity, thus my suggestion above. Of course, as more conditions get added, the clause may have to be rewritten again.

6by9 replied:
I was just retaining the format of the original term. I'll rework as requested.

6by9 committed:
staging/bcm2835-codec: Add VC-1 support.

Providing the relevant licence has been purchased, then Pi0-3
can decode VC-1.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
I've dropped the patch for enabling the decoder on STREAMON(OUTPUT) for now, but done the v4l2_m2m rework that was requested.
This has also gained a patch to enable VC1. It needs a firmware change to actually work reliably, but is harmless to merge this side first.

popcornmix commented:
I'm happy with this. Aspect ratio does work (needs an ffmpeg patch).
And VC1 is working.

popcornmix commented:
As mentioned on slack the VC1 patch doesn't support WMV9 (and the hardware/omplayer do).
So might be worth checking that before merging.

6by9 commented:
> As mentioned on slack the VC1 patch doesn't support WMV9 (and the hardware/omplayer do).
So might be worth checking that before merging.

Not going to happen in the short term. The codec is getting upset at quite a low level with that.
It needs a fair amount of analysis and bitstream parsing so that RIL can strip off the headers and then still pass the start of the stream into the codec for decoding. WMV is all special cases in RIL for that :-(

popcornmix commented:
Okay, I'm happy with PR then.
